### PR TITLE
HIVE-24936 - Fix file name parsing and copy file move.

### PR DESCRIPTION
**Pull request by harishjp**

### CODE REVIEW

**Comment by harishjp** [comment]
@kishendas, @prasanthj:  please help in reviewing this code.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/AbstractFileMergeOperator.java [hunk 1]
```diff
@@ -275,39 +275,33 @@ public void closeOp(boolean abort) throws HiveException {
...
         if (incompatFileSet != null && !incompatFileSet.isEmpty()) {
           for (Path incompatFile : incompatFileSet) {
-            // check if path conforms to Hive's file name convention. Hive expects filenames to be in specific format
+            // Hive expects filenames to be in specific format
             // like 000000_0, but "LOAD DATA" commands can let you add any files to any partitions/tables without
-            // renaming. This can cause MoveTask to remove files in some cases where MoveTask assumes the files are
-            // are generated by speculatively executed tasks.
+            // renaming.
+            // This can cause a few issues:
+            // MoveTask will remove files in some cases where MoveTask assumes the files are are generated by
```
Comment (kishendas) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/AbstractFileMergeOperator.java:285]: files are generated by

File: ql/src/java/org/apache/hadoop/hive/ql/exec/AbstractFileMergeOperator.java [hunk 2]
```diff
@@ -275,39 +275,33 @@ public void closeOp(boolean abort) throws HiveException {
...
-            }
+            // This can result in data loss in case of CONCATENATE/merging.
 
+            // If filename is consistent with XXXXXX_N and another task with same task-id runs after this move, then
+            // the same file name is used in the other task which will result in task failure and retry of task and
+            // subsequent removal of this file as duplicate.
+            // Example: if the file name is 000001_0 and another task runs with taskid 000001_0, it will fail to create
+            // the file and next attempt will create 000001_1, both the files will be considered as output of same task
+            // and only 000001_1 will be picked resulting it loss of existing file 000001_0.
+            final String destFileName = Utilities.getTaskId(jc) + Utilities.COPY_KEYWORD + 1;
```
Comment (kishendas) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/AbstractFileMergeOperator.java:299]: Will appending "+1" work in all cases ? 
What happens if concat partially succeeds and there is one more file with the same name, after the previous one is already moved.
If we run concat again, will it not result in data loss ?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/AbstractFileMergeOperator.java [hunk 3]
```diff
@@ -275,39 +275,33 @@ public void closeOp(boolean abort) throws HiveException {
...
-              destPath = destFilePath;
-            }
+            // This can result in data loss in case of CONCATENATE/merging.
 
+            // If filename is consistent with XXXXXX_N and another task with same task-id runs after this move, then
+            // the same file name is used in the other task which will result in task failure and retry of task and
+            // subsequent removal of this file as duplicate.
+            // Example: if the file name is 000001_0 and another task runs with taskid 000001_0, it will fail to create
+            // the file and next attempt will create 000001_1, both the files will be considered as output of same task
+            // and only 000001_1 will be picked resulting it loss of existing file 000001_0.
```
Comment (kishendas) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/AbstractFileMergeOperator.java:298]: resulting in loss of

File: ql/src/java/org/apache/hadoop/hive/ql/exec/ParsedOutputFileName.java [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+ * <p>
+ * All the components are here:
+ * tmp_(taskPrefix)00001_02_copy_1.zlib.gz
+ */
+public class ParsedOutputFileName {
+  private static final Pattern COPY_FILE_NAME_TO_TASK_ID_REGEX = Pattern.compile(
+      "^(.*?)?" + // any prefix
+      "(\\(.*\\))?" + // taskId prefix
+      "([0-9]+)" + // taskId
+      "(?:_([0-9]{1,6}))?" + // _<attemptId> (limited to 6 digits)
```
Comment (kishendas) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/ParsedOutputFileName.java:43]: Can attemptId be more than 6 digits in future ? Is it possible to refer to the regex for attemptId, from the place where its generated ?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java [hunk 5]
```diff
@@ -1093,40 +1093,69 @@ public static void rename(FileSystem fs, Path src, Path dst) throws IOException,
     }
   }
 
-  private static void moveFile(FileSystem fs, FileStatus file, Path dst) throws IOException,
+  private static void moveFileOrDir(FileSystem fs, FileStatus file, Path dst) throws IOException,
```
Comment (kishendas) [inline ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:1096]: Can we add tests to cover this method as well ?

File: ql/src/java/org/apache/hadoop/hive/ql/exec/AbstractFileMergeOperator.java (hunk 2 above)
Comment (harishjp) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/AbstractFileMergeOperator.java:299]: moveFile does not just move, if file already exists it keeps incrementing the copy index until it finds one which does not exist. So no data loss.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/ParsedOutputFileName.java (hunk 4 above)
Comment (harishjp) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/ParsedOutputFileName.java:43]: No idea, this was refactored from the existing pattern which assumed 6 digits, kept it same.

File: ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java (hunk 5 above)
Comment (harishjp) [reply ql/src/java/org/apache/hadoop/hive/ql/exec/Utilities.java:1096]: Test renameOrMove files calls this, and is the public API. Tests for that should cover cases here.

**Review by kishendas — APPROVED** [review]

**Pull request closed without merging**